### PR TITLE
Define shell function to automatically install license files

### PR DESCRIPTION
**Pull request by giordano**
This will be executed at the end of the build script.

Ref: https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/309

### CODE REVIEW

**Comment by giordano** [comment]
Aaaaargh, there is an error, wait!

**Comment by giordano** [comment]
Ok, it should be good now.  I had to turn `install_license` into a script because `find -exec install_license` can't run functions

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 1]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
 }
 
-# Function to install license files
-install_license () {
-    if [ $# -eq 0 ]; then
-        echo "Usage: install_license license_file1.txt [license_file2.md, license_file3.rtf, ...]" >&2
-        exit 1
+# Function to automatically install license files at the end of the build
+auto_install_license () {
+    if [[ -z "$(qfind "${prefix}/share/licenses/${SRC_NAME}" -mindepth 1)" ]]; then
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: I would rather just `-d ${prefix}/share/licenses/${SRC_NAME}`

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 2]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
+        # There are no licenses already installed, let's find 'em all
+
+        DIR="${WORKSPACE}/srcdir"
+        # Build the list of known names for license files
+        LICENSE_FILENAMES=()
+        for bname in COPYING COPYRIGHT LICENCE LICENSE ; do
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: I did not know that `licence` was a valid spelling.  :joy:

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 3]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
+            for extension in "" .md .txt; do
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: Add `.rtf`, I've seen that too.

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 4]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
+                # These are actually going to be options for `find`
+                LICENSE_FILENAMES+=(-iname "${bname}${extension}" -o)
+            done
+        done
+        # Remove the last OR
+        unset 'LICENSE_FILENAMES[${#LICENSE_FILENAMES[@]}-1]'
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: Ah, I did not know that's how you pop off the end of a bash array.  Neat!

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 5]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
+
+        # First round: look for license files in ${DIR}
+        qfind "${DIR}" -maxdepth 1 -type f \( "${LICENSE_FILENAMES[@]}" \) \
+              -exec install_license '{}' \;
+
+        if [[ -z "$(qfind "${prefix}/share/licenses/${SRC_NAME}" -mindepth 1)" ]]; then
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: This one should stay as `qfind`.

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 6]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
+            # We didn't install anything.  Let's see if there is a single subdir
+            # in ${DIR}, if so look for a license file inside it
+            dirs=$(qfind "${DIR}" -mindepth 1 -maxdepth 1 -type d -path "${DIR}/patches" -prune -o -print)
+            # This test is not safe if there are new lines in ${dirs}, but dealing
+            # with new lines here would require a bit of extra work, probably not
+            # worth the effort, as we can manually install the license if necessary
+            if [[ "$(echo "$dirs" | wc -l)" -eq 1 ]]; then
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: Agreed that this isn't worth the effort, but if you really did want to do this, you could do [something like this](https://stackoverflow.com/a/11131443/230778)

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh (hunk 1 above)
Comment (giordano) [reply 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: The idea of the test was to check if there are actually files inside the directory, not just if the dir is there.  Do you think that just checking the dir is sufficient?

Comment (staticfloat) [reply 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: I think the auto-install should fail out very aggressively; if someone tries to install a specific license and it doesn't work for some reason, the very fact that they tried (and created an empty directory) should be enough to scare off the auto-install routine, IMO.  We should only auto-install if absolutely no action to install a license has been taken.

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh (hunk 2 above)
Comment (giordano) [reply 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: It's used on this side of the pond :gb: :slightly_smiling_face: And I did find a file with this name in one of our packages, but cannot remember which one now.

File: 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh [hunk 7]
```diff
@@ -62,15 +62,40 @@ qfind() {
...
 }
 
-# Function to install license files
-install_license () {
-    if [ $# -eq 0 ]; then
-        echo "Usage: install_license license_file1.txt [license_file2.md, license_file3.rtf, ...]" >&2
-        exit 1
+# Function to automatically install license files at the end of the build
+auto_install_license () {
+    if [[ -d "${prefix}/share/licenses/${SRC_NAME}" ]]; then
```
Comment (staticfloat) [inline 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: Eh, my bad, this should be `! -d`.  :P

Comment (staticfloat) [reply 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: Or `[[ -d ... ]]; then return; fi` if you don't want to tab everything in.

Comment (giordano) [reply 0_RootFS/Rootfs/bundled/conf/profile.d/bb_utils.sh]: My bad too, didn't realise :see_no_evil: Fixed

**Pull request merged**